File: code_snippets/go/usecases/asynctasks/parallelizework/fanoutworker.go
package main

import (
	"context"
	restate "github.com/restatedev/sdk-go"
	"github.com/restatedev/sdk-go/server"
	"log/slog"
	"os"
)

// <start_here>
type FanOutWorker struct{}

func (FanOutWorker) Run(ctx restate.Context, task Task) (Result, error) {
	// Split the task in subtassk
	subtasks, err := split(task)
	if err != nil {
		return Result{}, err
	}

	// <mark_1>
	subtaskFutures := make([]restate.Selectable, 0, len(subtasks))
	for _, subtask := range subtasks {
		subtaskFutures = append(subtaskFutures,
			restate.Service[SubTaskResult](ctx, "FanOutWorker", "RunSubtask").RequestFuture(subtask))
	}
	// </mark_1>

	// <mark_2>
	selector := restate.Select(ctx, subtaskFutures...)
	// </mark_2>

	subResults := make([]SubTaskResult, 0, len(subtasks))
	for selector.Remaining() {
		// <mark_2>
		response, err := selector.Select().(restate.ResponseFuture[SubTaskResult]).Response()
		// </mark_2>
		if err != nil {
			return Result{}, err
		}
		subResults = append(subResults, response)
	}

	return aggregate(subResults)
}

func (FanOutWorker) RunSubtask(ctx restate.Context, subtask SubTask) (SubTaskResult, error) {
	// Processing logic goes here ...
	// Can be moved to a separate service or FaaS to scale independently
	return executeSubtask(ctx, subtask)
}

// <end_here>

func main() {
	server := server.NewRestate().
		Bind(restate.Reflect(FanOutWorker{}))

	if err := server.Start(context.Background(), ":9080"); err != nil {
		slog.Error("application exited unexpectedly", "err", err.Error())
		os.Exit(1)
	}
}

File: code_snippets/java/src/main/java/usecases/asynctasks/parallelize/FanOutWorker.java
package usecases.asynctasks.parallelize;

import static usecases.asynctasks.parallelize.utils.Utils.*;

import dev.restate.sdk.Context;
import dev.restate.sdk.DurableFuture;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Service;
import dev.restate.sdk.endpoint.Endpoint;
import dev.restate.sdk.http.vertx.RestateHttpServer;
import dev.restate.serde.TypeRef;
import java.util.ArrayList;
import java.util.List;
import usecases.asynctasks.parallelize.utils.*;

// <start_here>
@Service
public class FanOutWorker {

  @Handler
  public Result run(Context ctx, Task task) {

    // Split the task in subtasks
    List<SubTask> subTasks =
        ctx.run(
            new TypeRef<>() {},
            // break
            () -> split(task));

    List<DurableFuture<?>> resultFutures = new ArrayList<>();
    // <mark_1>
    for (SubTask subTask : subTasks) {
      resultFutures.add(FanOutWorkerClient.fromContext(ctx).runSubtask(subTask));
    }
    // </mark_1>

    // <mark_2>
    DurableFuture.all(resultFutures).await();

    var results = resultFutures.stream().map(future -> (SubTaskResult) future.await()).toList();
    // </mark_2>

    return aggregate(results);
  }

  @Handler
  public SubTaskResult runSubtask(Context ctx, SubTask subTask) {
    // Processing logic goes here ...
    // Can be moved to a separate service to scale independently
    return executeSubtask(ctx, subTask);
  }

  public static void main(String[] args) {
    RestateHttpServer.listen(Endpoint.bind(new FanOutWorker()));
  }
}
// <end_here>
